src/test/java/com/google/api/codegen/ProtoDocumentLinkTest.java

```diff
@@ -0,0 +1,99 @@
+/* Copyright 2016 Google Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen;
+
+import com.google.api.codegen.nodejs.JSDocCommentFixer;
+import com.google.api.codegen.py.PythonSphinxCommentFixer;
+import com.google.api.codegen.ruby.RDocCommentFixer;
+import com.google.common.truth.Truth;
+import java.util.regex.Matcher;
+import org.junit.Test;
+
+public class ProtoDocumentLinkTest {
+
+  @Test
+  public void testMatchProtoLink() {
+    Matcher m;
+    m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[Shelf][google.example.library.v1.Shelf]");
+    Truth.assertThat(m.find()).isTrue();
+    m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[Shelf][]");
+    Truth.assertThat(m.find()).isTrue();
+    m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[$Shelf][]");
+    Truth.assertThat(m.find()).isTrue();
+    m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[][abc]");
+    Truth.assertThat(m.find()).isFalse();
+    m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[A-Za-z_$][A-Za-z_$0-9]");
```

Comment:
The reason why [A-Za-z_$][A-Za-z_$0-9] is added as a test case is that it is a corner case causing the issue.